Add Subresource protections Article & Motivations for future defensive design article. (#16)
xsleaks/wiki: content/docs/defenses/design-protections/subresource-protections.md

+++
title = "Subresource Protections"
description = "Subresources Protections"
date = "2020-07-21"
category = "defenses"
menu = "main"
+++

### Random tokens

> One of the principles of protecting subresources is the same as protecting endpoints from [CSRF attacks](https://owasp.org/www-community/attacks/csrf). The difference from [CSRF protections](https://cheatsheetseries.owasp.org/cheatsheets/Cross-Site_Request_Forgery_Prevention_Cheat_Sheet.html) is that in the case of XS-Leaks, mostly GET requests are the ones worth protecting. To apply this protection, applications can append a (cryptographically strong) pseudorandom value, unique to each request/session, to make the URL of a subresource unpredictable to an attacker. The protection can be applied to the following types of subresources:

Reviewer: ```suggestion
One of the principles of protecting subresources is the same as protecting endpoints from [CSRF attacks](https://owasp.org/www-community/attacks/csrf). The difference from [CSRF protections](https://cheatsheetseries.owasp.org/cheatsheets/Cross-Site_Request_Forgery_Prevention_Cheat_Sheet.html) is that in the case of XS-Leaks, GET requests are the ones usually worth protecting. To apply this protection applications can append a (cryptographically strong) pseudorandom value, unique to each request/session, to make the URL of a subresource unpredictable to an attacker. The protection can be applied to the following types of subresources:
```

> - Authenticated subresources such as API endpoints or regular authenticated URLs. While pseudorandom values can be used in this case, security mitigations like [Same-Site Cookies](https://TODO) can be rather more effective.

Reviewer: ```suggestion
- Authenticated subresources such as API endpoints or regular authenticated URLs. While pseudorandom values can be used in this case, security mitigations like [Same-Site Cookies](https://TODO) can be cheaper to deploy and more effective.
```

> - Unauthenticated subresources such as images can use this protection to prevent some types of [Cache Probing Attacks](https://TODO). In this scenario, this protection can be highly effective.

Reviewer: But you basically remove the benefits from caches.

Author: Yes! but, maybe the website can still get the cached image if the random token used to construct the url is connected with the user session? (so the server should be able to systematically construct the same url, within a period of time?)

Author: check its hard to know wheter its worth protecting

### User Consent

> Some applications might ask for user consent to trigger a certain sensitive action. Facebook deploys this protection in some sensible search endpoints like `https://www.facebook.com/messages/?qa=UserMustConsent`, where a user musk press OK to advance with the search query. Since attackers can't surpass this verification, the page won't leak any special behavior.

Reviewer: I would also add the pre-redirect consent form.

Author: Right, and it seems common practise across multiple websites

Author: Added in c0f43de

Author: Add small reference that this might stop attackers from detecting navigations 18365a0

## Deployment

While this protection might work in some scenarios, it has some disadvantages:

- Hard to deploy as it requires substantial changes in the codebase. 
- It might break the desired behavior for the feature.

> - In the case of Random tokens, it will break bookmarks and other permanent references.

Reviewer: It might add friction to using the application (in the case of the extra consent)

Author: Added in c0f43de

+++
title = "Subresource Protections"
description = "Subresources Protections"
date = "2020-07-21"
category = "defenses"
menu = "main"
+++

### Random tokens

One of the principles of protecting subresources is the same as protecting endpoints from [CSRF attacks](https://owasp.org/www-community/attacks/csrf). The difference from [CSRF protections](https://cheatsheetseries.owasp.org/cheatsheets/Cross-Site_Request_Forgery_Prevention_Cheat_Sheet.html) is that in the case of XS-Leaks, GET requests are the ones usually worth protecting. To apply this protection applications can append a (cryptographically strong) pseudorandom value, unique to each request/session, to make the URL of a subresource unpredictable to an attacker. The protection can be applied to the following types of subresources:

- Authenticated subresources such as API endpoints or regular authenticated URLs. While pseudorandom values can be used in this case, security mitigations like [Same-Site Cookies](https://TODO) can be cheaper to deploy and more effective.

> - Unauthenticated subresources such as images can use this protection to prevent some types of [Cache Probing Attacks](https://TODO). In this scenario, this protection can be highly effective.

Reviewer: ```suggestion
- Unauthenticated subresources such as images can use this protection to prevent some types of [Cache Probing Attacks](https://TODO). In this scenario, this protection can be highly effective but it might hinder cache efficacy or be otherwise hard to deploy.
```
